app/src/main/java/ch/epfl/favo/common/DatabaseWrapper.java

```diff
@@ -45,22 +55,50 @@ public static String generateRandomId() {
     return sb.toString();
   }
 
-  public static void addDocument(String key, Map document, String collection)
-      throws RuntimeException {
-    DatabaseWrapper.getInstance()
-        .firestore
-        .collection(collection)
-        .document(key)
-        .set(document)
-        .addOnSuccessListener(aVoid -> {})
-        .addOnFailureListener(
-            e -> {
-              throw new RuntimeException(e);
-            });
+  static <T extends Document> void addDocument(T document, String collection) {
+    getCollectionReference(collection).document(document.getId()).set(document);
+  }
+
+  static <T extends Document> void removeDocument(String key, String collection) {
+    getCollectionReference(collection).document(key).delete();
+  }
+
+  static <T extends Document> void updateDocument(
+      String key, Map<String, Object> updates, String collection) {
+    getCollectionReference(collection).document(key).update(updates);
```

Comment:
We should replace getCollectionReference(collection) by a private attribute that we can inject and mock for testing

Reply:
The problem with that is `DatabaseWrapper` will no longer follow the singleton pattern if it stores state. The injection for testing is solved by the `DatabaseUpdater` class you created. `CollectionWrapper` is a non-singleton class that abstracts away from `DatabaseWrapper` and implements `DatabaseUpdater`.